plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -590,24 +595,88 @@ public boolean rename(Path src, Path dst)
     public boolean delete(Path path, boolean recursive)
             throws IOException
     {
-        try {
-            if (!directory(path)) {
-                return deleteObject(keyFromPath(path));
+        ObjectMetadata metadata = getS3ObjectMetadata(path);
+        if (metadata == null) {
+            // the bucket root does not exist
+            if (path.getName().isEmpty()) {
+                return false;
+            }
+
+            // check if this path is a directory
+            Iterator<LocatedFileStatus> iterator = listPath(path, OptionalInt.of(1), ListingMode.SHALLOW_ALL);
+            if (!iterator.hasNext()) {
+                return false;
             }
         }
-        catch (FileNotFoundException e) {
-            return false;
+        else {
+            // the bucket root requires special handling
+            if (!path.getName().isEmpty()) {
+                MediaType metadataContentType = MediaType.parse(metadata.getContentType());
+                if (!metadataContentType.is(DIRECTORY_MEDIA_TYPE)) {
+                    if (!listPath(path, OptionalInt.of(1), ListingMode.SHALLOW_ALL).hasNext()) {
+                        // delete the object
+                        return deleteObject(keyFromPath(path));
+                    }
+
+                    // Delete the object which has a key corresponding to the prefix to be deleted
+                    deleteObject(keyFromPath(path));
+                }
+            }
         }
 
         if (!recursive) {
             throw new IOException("Directory " + path + " is not empty");
         }
 
-        for (FileStatus file : listStatus(path)) {
-            delete(file.getPath(), true);
+        String bucketName = getBucketName(uri);
+        Iterator<LocatedFileStatus> listings = listPath(path, OptionalInt.empty(), ListingMode.RECURSIVE_FILES_ONLY);
```

Comment:
`ListingMode.RECURSIVE_FILES_ONLY` will filter out "directories" that have corresponding S3 keys/objects, which we do want to delete when performing a recursive deletion. See [my other comment](https://github.com/trinodb/trino/pull/13974#discussion_r964867348) about this.